### Heroes Of Pymoli Data Analysis
* Of the 1163 active players, the vast majority are male (84%). There also exists, a smaller, but notable proportion of female players (14%).

* Our peak age demographic falls between 20-24 (44.8%) with secondary groups falling between 15-19 (18.60%) and 25-29 (13.4%).  
-----

### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [38]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# File to Load (Remember to Change These)
file_to_load = "./purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)
purchase_data.head()

#purchase_data.columns
#Index(['Purchase ID', 'SN', 'Age', 'Gender', 'Item ID', 'Item Name', 'Price'], dtype='object')

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count

* Display the total number of players


In [39]:
#df_players = purchase_data.loc[:,"SN"]
#df_players.head()

#Find the toal number of players, by ensuring we are selecting only the unique ones
unique_players = len(purchase_data["SN"].unique())

# Now put this information into a Pandas Dataframe for printout

overview_df = pd.DataFrame({
        "Total #Players":[unique_players]
})
overview_df


Total #Players
0             576

#  76% of the players in this dataset have purchased at least 1 or more items
This dataset of 576 is a subset of the total as described in the Homework Description

576 players are unique (out of a listing of 760 players) and each have purchased at least 1 item

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [40]:
# Find the unique items, total revenue, the # of purchases, and the average purchase price

unique_items = len(purchase_data["Item ID"].unique())

total_revenue = purchase_data["Price"].sum()
numof_purchases = len(purchase_data["SN"])

average_price = total_revenue / numof_purchases

purchase_analysis_df = pd.DataFrame({
        "Unique items":[unique_items],
        "Total #Purchases":[numof_purchases],
        "Average Price":[average_price],
        "Total Revenue":[total_revenue]
})

# Format
purchase_analysis_df["Average Price"] = purchase_analysis_df["Average Price"].astype(float).map("${:,.2f}".format)
purchase_analysis_df["Total Revenue"] = purchase_analysis_df["Total Revenue"].astype(float).map("${:,.2f}".format)

purchase_analysis_df

Unique items  Total #Purchases Average Price Total Revenue
0           183               780         $3.05     $2,379.77

# 76% of the items have been purchased more than once

The number of unique items is ~24%  of the total purchases  (183 / 780) .  
This means that 597 items (76%) have been purchased at least more than once

The total revenue of purchases is ~$2400 for the sample data we have
The average price of an item is ~$3.05

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [41]:
# Calculating Player demographics based on gender.  

male_players = purchase_data.loc[purchase_data["Gender"] == "Male"]
num_unique_male = len(male_players["SN"].unique())

female_players =  purchase_data.loc[purchase_data["Gender"] == "Female"]
num_unique_female = len(female_players["SN"].unique())


other_players = purchase_data.loc[purchase_data["Gender"] == "Other / Non-Disclosed"]
num_unique_other = len(other_players["SN"].unique())

                    
#unique_male, unique_female, unique_other

percent_male = round(num_unique_male / unique_players * 100,2)
percent_female = round(num_unique_female / unique_players * 100,2)
percent_other = round(num_unique_other / unique_players * 100,2)

# Display in a dataframe
demographics_df = pd.DataFrame({
        "Gender":["Male","Female", "Other/Non-Disclosed"],
        "Total Count":[num_unique_male, num_unique_female, num_unique_other],
        "% of Players":[percent_male, percent_female, percent_other]
})
demo_print = demographics_df.set_index("Gender")


demo_print

Total Count  % of Players
Gender                                        
Male                         484         84.03
Female                        81         14.06
Other/Non-Disclosed           11          1.91

# 84% of Pymoli players are male,  14% are female

This is already a conclusions given at the top of the assignment, and this dataset (which is a subset of the total), has the same distribution as a function of gender


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [9]:
# Purchasing Analysis by Gender
# Note:  I recognize this is a brute force way to create the dataframe relative to what I learned on later exercises

#purchases_byGender = purchase_data[["SN", "Age", "Gender","Item ID","Item Name", "Price"]]
#purchases_byGender.head()
#purchases_byGender.count()

# Calculate puchases for Men
male_items_count = len(male_players["Item ID"])
male_total_spent = male_players["Price"].sum()                               
male_ave_price = male_total_spent / male_items_count
male_ave_price_perperson = male_total_spent / num_unique_male
                      
# Calculate puchases for Women
female_items_count = len(female_players["Item ID"])
female_total_spent = female_players["Price"].sum()                         
female_ave_price = female_total_spent / female_items_count
female_ave_price_perperson = female_total_spent / num_unique_female
                            
 # Calculate puchases for Non-disclosed
other_items_count = len(other_players["Item ID"])
other_total_spent = other_players["Price"].sum()
other_ave_price = other_total_spent / other_items_count
other_ave_price_perperson = other_total_spent / num_unique_other

# Put Results into a Dataframe
purchase_bygender_df = pd.DataFrame({
        "Gender":["Male","Female", "Other/Non-Disclosed"],
        "Total #Purchases":[male_items_count, female_items_count, other_items_count],
        "Total Spent($)":[male_total_spent, female_total_spent, other_total_spent],
        "Ave Purchase Price($)":[male_ave_price, female_ave_price, other_ave_price],
        "Ave Spend Per Person($)":[male_ave_price_perperson, female_ave_price_perperson, other_ave_price_perperson]
})

purchase_bygender_df.head()

# Formatting of the Display
purchase_bygender_df["Total Spent($)"] = purchase_bygender_df["Total Spent($)"].map("${:.2f}".format)
purchase_bygender_df["Ave Purchase Price($)"] = purchase_bygender_df["Ave Purchase Price($)"].map("${:.2f}".format)
purchase_bygender_df["Ave Spend Per Person($)"] = purchase_bygender_df["Ave Spend Per Person($)"].map("${:.2f}".format)

purchase_bygender_print = purchase_bygender_df.set_index("Gender")
purchase_bygender_print





Total #Purchases Total Spent($) Ave Purchase Price($)  \
Gender                                                                       
Male                              652       $1967.64                 $3.02   
Female                            113        $361.94                 $3.20   
Other/Non-Disclosed                15         $50.19                 $3.35   

                    Ave Spend Per Person($)  
Gender                                       
Male                                  $4.07  
Female                                $4.47  
Other/Non-Disclosed                   $4.56

#  Player spending is similar by gender
Males (84% of the player population) make ~83.5% of the purchases
Females (14% of the player population) make ~14% of the purchases, and spend ~10% more per person than males, and purchase higher priced items on average (6% higher spend)

Non-Disclosed (~2% of the population) make ~2% of the purchases, with slightly higher spend per person (11%), and buying higher priced items

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [36]:

#Create a unique player list with ages

age_df = purchase_data[["SN", "Age"]]
age_unique = age_df.drop_duplicates()
age_unique = age_unique[["Age"]]


# Create bins and labels for the bins for the values to be placed

bins = [0,9,14,19,24,29,34,39,100]
group_labels = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]

# Display the age group and the number of players per bin
age_unique["Age Group"] = pd.cut(age_unique["Age"], bins, labels=group_labels)
age_group = age_unique.groupby("Age Group").count()

# Add a column for the % of players and format the display
age_group["% of Players"] = round(age_group["Age"] /unique_players *100,2)
age_group["% of Players"] = age_group["% of Players"].map("{:.1f}%".format)


# Correct the column name 
age_group = age_group.rename(
    columns={"Age": "# of Players"})

age_group.head(8)


# of Players % of Players
Age Group                           
<10                  17         3.0%
10-14                22         3.8%
15-19               107        18.6%
20-24               258        44.8%
25-29                77        13.4%
30-34                52         9.0%
35-39                31         5.4%
40+                  12         2.1%

#  45% of the players are age 20-24,  76% players age 15-59
This was already stated in the Homework description,
Which means that this subset of data fits within the overall statistics

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [11]:
# Use the existing bins and labels-
# 1. Calculate the number of purchases per bin
# 2. Calculate the avg. purchase price per bin
# 3. Calculate the average purchase total per person,  in each bin
# 4. join the tables together based on Age Group

# re-using the bin and label references from above
#bins = [0,9,14,19,24,29,34,39,100]
#group_labels = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]

purchase_data.head()


Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [42]:
# 1 Number of purchases, by age group

purchase_byage = purchase_data.loc[:,["Purchase ID", "Age"]]
purchase_byage["Age Group"] = pd.cut(purchase_byage["Age"], bins, labels=group_labels)

purchase_group = purchase_byage.groupby("Age Group").count()

# Correct the column name 
purchase_group = purchase_group.rename(
    columns={"Purchase ID": "#Items Purchased"})

del purchase_group["Age"]
purchase_group.head(8)

#Items Purchased
Age Group                  
<10                      23
10-14                    28
15-19                   136
20-24                   365
25-29                   101
30-34                    73
35-39                    41
40+                      13

In [43]:
#2 Mean Price of objects purchased, by Age Group
mean_price = purchase_data.loc[:,["Age","Price"]]
mean_price["Age Group"] = pd.cut(mean_price["Age"], bins, labels = group_labels)
mean_price = mean_price.groupby("Age Group").mean()


#Rename columns and remove unnecessary columns
mean_price["Price"]=mean_price["Price"].map("${:.2f}".format)
mean_price = mean_price.rename(columns={"Price": "Average Price"})
del mean_price["Age"]

mean_price.head(8)


Average Price
Age Group              
<10               $3.35
10-14             $2.96
15-19             $3.04
20-24             $3.05
25-29             $2.90
30-34             $2.93
35-39             $3.60
40+               $2.94

In [44]:
#3 total spend per bin and per person
# Total spend per bin
total_spend = purchase_data.loc[:,["Age","Price"]]
total_spend["Age Group"] = pd.cut(total_spend["Age"], bins, labels = group_labels)
total_spend = total_spend.groupby("Age Group").sum()

# Total spend per person (pp = per person),  first by merging to one table
merge1 = pd.merge(age_group,total_spend, on="Age Group", how="outer")
merge1.head()

total_spend_pp = merge1["Price"] / merge1["# of Players"]
total_spend["Total Spend Per Player"] = total_spend_pp

# Update format and remove unnecessary columns
total_spend["Total Spend Per Player"]=total_spend["Total Spend Per Player"].map("${:.2f}".format)
total_spend["Price"]=total_spend["Price"].map("${:.2f}".format)
total_spend = total_spend.rename(columns={"Price": "Total Spend"})
del total_spend["Age"]


total_spend.head(8)


Total Spend Total Spend Per Player
Age Group                                   
<10            $77.13                  $4.54
10-14          $82.78                  $3.76
15-19         $412.89                  $3.86
20-24        $1114.06                  $4.32
25-29         $293.00                  $3.81
30-34         $214.00                  $4.12
35-39         $147.67                  $4.76
40+            $38.24                  $3.19

In [45]:
# Merge all the data together,  including the Age Demographics

merge_table = pd.merge(age_group, purchase_group, on="Age Group", how="outer")
merge_table = pd.merge(merge_table, mean_price, on="Age Group", how="outer")
merge_table = pd.merge(merge_table, total_spend, on="Age Group", how="outer")
merge_table

# of Players % of Players  #Items Purchased Average Price  \
Age Group                                                              
<10                  17         3.0%                23         $3.35   
10-14                22         3.8%                28         $2.96   
15-19               107        18.6%               136         $3.04   
20-24               258        44.8%               365         $3.05   
25-29                77        13.4%               101         $2.90   
30-34                52         9.0%                73         $2.93   
35-39                31         5.4%                41         $3.60   
40+                  12         2.1%                13         $2.94   

          Total Spend Total Spend Per Player  
Age Group                                     
<10            $77.13                  $4.54  
10-14          $82.78                  $3.76  
15-19         $412.89                  $3.86  
20-24        $1114.06                  $4.32  
25-29         $293.00                  $3.81  
30-34         $214.00                  $4.12  
35-39         $147.67                  $4.76  
40+            $38.24                  $3.19

#       20-24 year olds (45% of the players) make 63% of the purchases, with 47% of the Spend

Every other category make similar or slightly fewer purchases per person.
Total spending amount per group is very consistent with the number of players in the group.



## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [46]:
#  Reference the original data provided 
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [47]:
#Find the top 5 spenders, # purchases, their Total spend and the average spend per item

# Clean the list to those players that purchased more than one item
# https://stackoverflow.com/questions/14657241/how-do-i-get-a-list-of-all-the-duplicate-items-using-pandas-in-python

multi_purchases =purchase_data[purchase_data.duplicated(subset="SN", keep = False)]

# get the list of spenders from max to min, put it into a dataframe
spenders = multi_purchases["SN"].value_counts()

spenders_df = pd.DataFrame({
        "# Purchases":spenders
})

spenders_df.head()

# Purchases
Lisosia93              5
Iral74                 4
Idastidru52            4
Hada39                 3
Zontibe81              3

In [48]:
# get the total spend by person

total_spend = multi_purchases.loc[:,["SN", "Price"]]
total_spend = total_spend.groupby("SN").sum()

#total_spend["Total Spend Per Player"] = pd.to_numeric(total_spend["Total Spend Per Player"])

# The formatting code below seems toO MAKE PRICE AN OBJECT RATHER THAN FLOAT64, so that it can't be used for computation
# Hold off on formatting to the final table
#total_spend["Price"]=total_spend["Price"].map("${:.2f}".format) 

total_spend = total_spend.rename(columns={"Price": "Total Spend Per Player"})

total_spend.head()


Total Spend Per Player
SN                                   
Aelastirin39                     7.29
Aelin32                          8.98
Aelly27                          6.79
Aelollo59                        5.63
Aerithllora36                    8.64

In [49]:
# Merge the # purchases and the mean price
#merge_table = pd.merge(total_spend, spenders_df, on="SN", how="inner")
join_result = total_spend.join(spenders_df, on="SN", how="inner")

join_result = join_result.sort_values(by ="Total Spend Per Player",ascending = False)

ave_purchase = join_result["Total Spend Per Player"] / join_result["# Purchases"]
join_result["Ave Purchase Cost $"] = round(ave_purchase,2)

#calculation for Only the top x players   - This section of code gives a pink warning box. (Why?) Skip for now.
#xnumofplayers = 5
#top_x_spenders = xnumofplayers - len(join_result["# Purchases"])
#topspenders_df = join_result[:top_x_spenders]
#topspenders_df["Total Spend Per Player"] = topspenders_df["Total Spend Per Player"].map("${:.2f}".format) 
#topspenders["Ave Purchase Cost $"] = topspenders["Ave Purchase Cost $"].map("${:.2f}".format)
#topspenders_df

# Format results -- no pink warning box
join_result["Total Spend Per Player"] = join_result["Total Spend Per Player"].map("${:.2f}".format) 
join_result["Ave Purchase Cost $"] = join_result["Ave Purchase Cost $"].map("${:.2f}".format)
join_result.head()

Total Spend Per Player  # Purchases Ave Purchase Cost $
SN                                                                 
Lisosia93                   $18.96            5               $3.79
Idastidru52                 $15.45            4               $3.86
Chamjask73                  $13.83            3               $4.61
Iral74                      $13.62            4               $3.40
Iskadarya95                 $13.10            3               $4.37

# The biggest spenders spent 13-19 dollars per player and purchased 3-5 items

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [50]:
# Analyse popular products, by # of purchases, and by total sales

popular = purchase_data.loc[:,["Item ID", "Item Name", "Price"]]
num_of_purchases = popular.groupby(["Item ID", "Item Name"]).count()
num_of_purchases = num_of_purchases.rename(columns={"Price":"# Purchased"})

popular_group = popular.groupby(["Item ID", "Item Name"])
popular_df = popular_group["Price"].agg([np.sum,np.mean ])

popular_df = popular_df.rename(columns={"sum" : "Total Sales $", "mean" : "Ave Price $"})

popular_table = num_of_purchases.join(popular_df, on=['Item ID', 'Item Name'])

most_popular = popular_table.sort_values(by = "# Purchased", ascending = False)

most_popular["Total Sales $"] = popular_table["Total Sales $"].map("${:.2f}".format) 
most_popular["Ave Price $"] = popular_table["Ave Price $"].map("${:.2f}".format)

most_popular.head(15)
# calculates the most popular by only looking at the #purchased

# Purchased  \
Item ID Item Name                                                   
178     Oathbreaker, Last Hope of the Breaking Storm           12   
145     Fiery Glass Crusader                                    9   
108     Extraction, Quickblade Of Trembling Hands               9   
82      Nirvana                                                 9   
19      Pursuit, Cudgel of Necromancy                           8   
103     Singed Scalpel                                          8   
75      Brutality Ivory Warmace                                 8   
72      Winter's Bite                                           8   
60      Wolf                                                    8   
59      Lightning, Etcher of the King                           8   
37      Shadow Strike, Glory of Ending Hope                     8   
34      Retribution Axe                                         8   
92      Final Critic                                            8   
53      Vengeance Cleaver                                       7   
110     Suspension                                              7   

                                                     Total Sales $ Ave Price $  
Item ID Item Name                                                               
178     Oathbreaker, Last Hope of the Breaking Storm        $50.76       $4.23  
145     Fiery Glass Crusader                                $41.22       $4.58  
108     Extraction, Quickblade Of Trembling Hands           $31.77       $3.53  
82      Nirvana                                             $44.10       $4.90  
19      Pursuit, Cudgel of Necromancy                        $8.16       $1.02  
103     Singed Scalpel                                      $34.80       $4.35  
75      Brutality Ivory Warmace                             $19.36       $2.42  
72      Winter's Bite                                       $30.16       $3.77  
60      Wolf                                                $28.32       $3.54  
59      Lightning, Etcher of the King                       $33.84       $4.23  
37      Shadow Strike, Glory of Ending Hope                 $25.28       $3.16  
34      Retribution Axe                                     $17.76       $2.22  
92      Final Critic                                        $39.04       $4.88  
53      Vengeance Cleaver                                   $14.35       $2.05  
110     Suspension                                          $10.08       $1.44

In [51]:
# Sort on the most popular by # purchased -and- then sort by highest sales
# Because the most popular items are a combination of how many purchased and whether they are willing to pay a higher price

most_popular_good_sales = popular_table.sort_values(by = ["# Purchased", "Total Sales $"], ascending = False)

most_popular_good_sales["Total Sales $"] = most_popular_good_sales["Total Sales $"].map("${:.2f}".format) 
most_popular_good_sales["Ave Price $"] = most_popular_good_sales["Ave Price $"].map("${:.2f}".format)
most_popular_good_sales



# Purchased  \
Item ID Item Name                                                   
178     Oathbreaker, Last Hope of the Breaking Storm           12   
82      Nirvana                                                 9   
145     Fiery Glass Crusader                                    9   
108     Extraction, Quickblade Of Trembling Hands               9   
92      Final Critic                                            8   
103     Singed Scalpel                                          8   
59      Lightning, Etcher of the King                           8   
72      Winter's Bite                                           8   
60      Wolf                                                    8   
37      Shadow Strike, Glory of Ending Hope                     8   
75      Brutality Ivory Warmace                                 8   
34      Retribution Axe                                         8   
19      Pursuit, Cudgel of Necromancy                           8   
78      Glimmer, Ender of the Moon                              7   
141     Persuasion                                              7   
159     Oathbreaker, Spellblade of Trials                       7   
53      Vengeance Cleaver                                       7   
117     Heartstriker, Legacy of the Light                       7   
85      Malificent Bag                                          7   
164     Exiled Doomblade                                        7   
71      Demise                                                  7   
110     Suspension                                              7   
7       Thorn, Satchel of Dark Souls                            7   
179     Wolf, Promise of the Moonwalker                         6   
160     Azurewrath                                              6   
25      Hero Cane                                               6   
136     Ghastly Adamantite Protector                            6   
102     Avenger                                                 6   
96      Blood-Forged Skeletal Spine                             6   
120     Agatha                                                  6   
...                                                           ...   
81      Dreamkiss                                               2   
43      Foul Edge                                               2   
132     Persuasion                                              2   
56      Foul Titanium Battle Axe                                2   
176     Relentless Iron Skewer                                  2   
14      Possessed Core                                          2   
48      Rage, Legacy of the Lone Victor                         2   
158     Darkheart, Butcher of the Champion                      2   
30      Stormcaller                                             2   
177     Winterthorn, Defender of Shifting Worlds                2   
115     Spectral Diamond Doomblade                              2   
69      Frenzy, Defender of the Harvest                         2   
31      Trickster                                               2   
168     Sun Strike, Jaws of Twisted Visions                     2   
33      Curved Axe                                              2   
26      Unholy Wand                                             2   
28      Flux, Destroyer of Due Diligence                        2   
125     Whistling Mithril Warblade                              2   
51      Endbringer                                              1   
134     Undead Crusader                                         1   
91      Celeste                                                 1   
47      Alpha, Reach of Ending Hope                             1   
180     Stormcaller                                             1   
27      Riddle, Tribute of Ended Dreams                         1   
90      Betrayer                                                1   
118     Ghost Reaver, Longsword of Magic            

In [52]:
# Compare to the result if "popularity"  is defined by by higher total sales (sort by only the Total Sales)

most_popular_bysales = popular_table.sort_values(by = ["Total Sales $"], ascending = False)

most_popular_bysales["Total Sales $"] = most_popular_bysales["Total Sales $"].map("${:.2f}".format) 
most_popular_bysales["Ave Price $"] = most_popular_bysales["Ave Price $"].map("${:.2f}".format)

most_popular_bysales.head()

# Purchased  \
Item ID Item Name                                                   
178     Oathbreaker, Last Hope of the Breaking Storm           12   
82      Nirvana                                                 9   
145     Fiery Glass Crusader                                    9   
92      Final Critic                                            8   
103     Singed Scalpel                                          8   

                                                     Total Sales $ Ave Price $  
Item ID Item Name                                                               
178     Oathbreaker, Last Hope of the Breaking Storm        $50.76       $4.23  
82      Nirvana                                             $44.10       $4.90  
145     Fiery Glass Crusader                                $41.22       $4.58  
92      Final Critic                                        $39.04       $4.88  
103     Singed Scalpel                                      $34.80       $4.35

#   The Oathbreaker is the most popular item, in terms of number sold

12 items sold,  which is 30% more than the next most popular items.

The interesting question is whether 

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [53]:
# Note Total Purchase Value = Total Sales  of the item
#List of highest total sales

Profit_table = popular_table.sort_values(by = "Total Sales $", ascending = False)

Profit_table["Total Sales $"] = Profit_table["Total Sales $"].map("${:.2f}".format) 
Profit_table["Ave Price $"] = Profit_table["Ave Price $"].map("${:.2f}".format)

Profit_table.head()

#List of highest total sales

# Purchased  \
Item ID Item Name                                                   
178     Oathbreaker, Last Hope of the Breaking Storm           12   
82      Nirvana                                                 9   
145     Fiery Glass Crusader                                    9   
92      Final Critic                                            8   
103     Singed Scalpel                                          8   

                                                     Total Sales $ Ave Price $  
Item ID Item Name                                                               
178     Oathbreaker, Last Hope of the Breaking Storm        $50.76       $4.23  
82      Nirvana                                             $44.10       $4.90  
145     Fiery Glass Crusader                                $41.22       $4.58  
92      Final Critic                                        $39.04       $4.88  
103     Singed Scalpel                                      $34.80       $4.35

# The Oathbreaker was also the most profitable,  due to the higher number of sales (12)  and higher pricing  (4.23 / item)

  The average sales price for Oathbreaker was not the highest price of all items, but still in the top 50 of prices. See dataframe below.
  
  The Oathbreaker appears to deliver a lot of benefits for the price.
  
  This suggests that if more profit is desired, it's possible to raise the pricing of this item.
  
  Alternatively,  if some of the other high priced items are not selling well due to lack of benefits,  reduced pricing may be helpful to drive sales up of other items.



In [74]:
# Pricing Table
Pricing_table = popular_table.sort_values(by = ["Ave Price $"], ascending = False)

Pricing_table["Total Sales $"] = Pricing_table["Total Sales $"].map("${:.2f}".format) 
Pricing_table["Ave Price $"] = Pricing_table["Ave Price $"].map("${:.2f}".format)

Pricing_table.head(50)

# Purchased  \
Item ID Item Name                                                   
63      Stormfury Mace                                          2   
139     Mercy, Katana of Dismay                                 5   
173     Stormfury Longsword                                     2   
147     Hellreaver, Heirloom of Inception                       3   
128     Blazeguard, Reach of Eternity                           5   
82      Nirvana                                                 9   
111     Misery's End                                            4   
92      Final Critic                                            8   
165     Bone Crushing Silver Skewer                             4   
52      Hatred                                                  4   
64      Fusion Pummel                                           3   
157     Spada, Etcher of Hatred                                 4   
109     Downfall, Scalpel Of The Emperor                        3   
41      Orbit                                                   3   
95      Singed Onyx Warscythe                                   3   
138     Peacekeeper, Wit of Dark Magic                          5   
51      Endbringer                                              1   
18      Torchlight, Bond of Storms                              2   
89      Blazefury, Protector of Delusions                       3   
57      Despair, Favor of Due Diligence                         4   
50      Dawn                                                    4   
145     Fiery Glass Crusader                                    9   
134     Undead Crusader                                         1   
175     Woeful Adamantite Claymore                              3   
179     Wolf, Promise of the Moonwalker                         6   
87      Deluge, Edge of the West                                5   
66      Victor Iron Spikes                                      4   
154     Feral Katana                                            4   
78      Glimmer, Ender of the Moon                              7   
160     Azurewrath                                              6   
25      Hero Cane                                               6   
103     Singed Scalpel                                          8   
119     Stormbringer, Dark Blade of Ending Misery               3   
68      Storm-Weaver, Slayer of Inception                       4   
65      Conqueror Adamantite Mace                               5   
178     Oathbreaker, Last Hope of the Breaking Storm           12   
59      Lightning, Etcher of the King                           8   
101     Final Critic                                            5   
116     Renewed Skeletal Katana                                 4   
74      Yearning Crusher                                        3   
91      Celeste                                                 1   
58      Freak's Bite, Favor of Holy Might                       4   
133     Faith's Scimitar                                        5   
5       Putrid Fan                                              4   
113     Solitude's Reaver                                       4   
79      Alpha, Oath of Zeal                                     3   
148     Warmonger, Gift of Suffering's End                      3   
182     Toothpick                                               3   
77      Piety, Guardian of Riddles                              4   
170     Shadowsteel                                             4   

                                                     Total Sales $ Ave Price $  
Item ID Item Name                                                               
63      Stormfury Mace                                       $9.98       $4.99  
139     Mercy, Katana of Dismay                             $24.70       $4.94  
173     Stormfury Longsword                                  $9.86       $4.93  
147     Hellreaver, Heirloom of Inception                   